// @ay/ui/dist/dividing/dividing.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ViewEncapsulation, forwardRef } from '@angular/core';
import { AnchorComponent } from "../scroll/anchor.component";
var DividingComponent = (function (_super) {
    __extends(DividingComponent, _super);
    function DividingComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DividingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ay-dividing',
                    template: "\n    <div class=\"border left\" [ngClass]=\"{ empty: ref.innerHTML.length == 0}\"></div><div class=\"content\" #ref=\"\"><ng-content></ng-content></div><div class=\"border right\" [ngClass]=\"{ empty: ref.innerHTML.length == 0}\"></div>\n  ",
                    styles: ["\n    ay-dividing {\n      display: block;\n      margin: 1.5em 0;\n      color: #999;\n      height: 1em;\n      display: flex;\n      flex-direction: row;\n      text-align: center; }\n      ay-dividing .border {\n        flex: 1;\n        border-top: 1px solid #CCC;\n        margin: .5em;\n        margin-bottom: 0;\n        height: .5em; }\n      ay-dividing .content {\n        line-height: 1em;\n        vertical-align: top; }\n      ay-dividing .left {\n        margin-right: 1em; }\n      ay-dividing .right {\n        margin-left: 1em; }\n      ay-dividing .left.empty {\n        margin-right: 0em; }\n      ay-dividing .right.empty {\n        margin-left: 0em; }\n\n    .flex-row > ay-dividing {\n      flex-direction: column;\n      margin: 0 1.5em;\n      height: inherit; }\n      .flex-row > ay-dividing .border {\n        border: none;\n        border-right: 1px solid #CCC;\n        margin: 0; }\n      .flex-row > ay-dividing .content {\n        width: 1em; }\n      .flex-row > ay-dividing .left {\n        margin-bottom: 1em;\n        margin-right: .5em; }\n      .flex-row > ay-dividing .right {\n        margin-top: 1em;\n        margin-right: .5em; }\n      .flex-row > ay-dividing .left.empty {\n        margin-bottom: 0em;\n        margin-right: .5em; }\n      .flex-row > ay-dividing .right.empty {\n        margin-top: 0em;\n        margin-right: .5em; }\n  "],
                    encapsulation: ViewEncapsulation.None,
                    providers: [{
                            provide: AnchorComponent,
                            useExisting: forwardRef(function () { return DividingComponent; })
                        }]
                },] },
    ];
    /** @nocollapse */
    DividingComponent.ctorParameters = function () { return []; };
    return DividingComponent;
}(AnchorComponent));
export { DividingComponent };
