// @ay/ui/dist/ay.module.js
import { NgModule } from "@angular/core";
import { DropdownModule, DropdownComponent } from "./dropdown/dropdown.module";
import { ScrollTopModule, ScrollTopDirective } from "./scroll-top/scroll-top.module";
import { ScrollLeftModule, ScrollLeftDirective } from "./scroll-left/scroll-left.module";
import { ScrollModule, ScrollComponent } from "./scroll/scroll.module";
import { DateModule, DateComponent } from "./date/date.module";
import { TimeModule, TimeComponent } from "./time/time.module";
import { DateTimeModule, DateTimeComponent } from "./datetime/datetime.module";
import { AlertModule, AlertService } from "./alert/alert.module";
import { ColorModule, ColorComponent } from "./color/color.module";
import { SliderModule, SliderComponent } from "./slider/slider.module";
import { ToggleModule, ToggleComponent } from "./toggle/toggle.module";
import { CheckboxModule, CheckboxComponent } from "./checkbox/checkbox.module";
import { AutocompleteModule, AutocompleteComponent } from "./autocomplete/autocomplete.module";
import { ButtonModule, ButtonComponent } from "./button/button.module";
import { TableModule, TableComponent } from "./table/table.module";
import { ListModule, ListComponent } from "./list/list.module";
import { TooltipModule, TooltipComponent } from "./tooltip/tooltip.module";
import { InputModule, InputComponent } from "./input/input.module";
import { DividingModule, DividingComponent } from "./dividing/dividing.module";
import { NotifyModule, NotifyService, Notify } from "./notify/notify.module";
import { LoadingModule, LoadingComponent, LoadingService } from "./loading/loading.module";
import { ShowModule, ShowDirective } from "./show/show.module";
import { FormatModule, FormatPipe } from "./format/format.module";
import { BrandModule, BrandComponent } from "./brand/brand.module";
import { PageModule, PageColmponent } from "./page/page.module";
import { ChipsModule, ChipsComponent } from "./chips/chips.module";
import { CardsComponent, CardsModule } from "./cards/cards.module";
import { TextDateComponent, TextDateModule } from "./textDate/textDate.module";
import { SelectableModule } from "./selectable/selectable.module";
import { DndModule, DragDropService, DragDropSortableService, DragDropConfig } from 'ng2-dnd';
import { SelectableService } from "./selectable/selectable.service";
export { SelectableService, DragDropConfig, DragDropService, DragDropSortableService, DropdownComponent, ScrollTopDirective, ScrollLeftDirective, ScrollComponent, DateComponent, TimeComponent, DateTimeComponent, AlertService, ColorComponent, SliderComponent, ToggleComponent, CheckboxComponent, AutocompleteComponent, ButtonComponent, TableComponent, ListComponent, TooltipComponent, InputComponent, DividingComponent, NotifyService, Notify, LoadingComponent, LoadingService, ShowDirective, FormatPipe, BrandComponent, PageColmponent, ChipsComponent, CardsComponent, TextDateComponent, };
var AY_MODULES = [
    SelectableModule,
    DropdownModule,
    ScrollTopModule,
    ScrollLeftModule,
    ScrollModule,
    DateModule,
    TimeModule,
    DateTimeModule,
    AlertModule,
    ColorModule,
    SliderModule,
    ToggleModule,
    CheckboxModule,
    AutocompleteModule,
    ButtonModule,
    TableModule,
    ListModule,
    TooltipModule,
    InputModule,
    DividingModule,
    DndModule,
    NotifyModule,
    LoadingModule,
    ShowModule,
    FormatModule,
    BrandModule,
    PageModule,
    ChipsModule,
    CardsModule,
    TextDateModule,
];
var AyModule = (function () {
    function AyModule() {
    }
    AyModule.forRoot = function () {
        return {
            ngModule: AyModule,
        };
    };
    AyModule.decorators = [
        { type: NgModule, args: [{
                    imports: AY_MODULES,
                    exports: AY_MODULES,
                    declarations: [],
                    providers: [DragDropService, DragDropSortableService, DragDropConfig]
                },] },
    ];
    /** @nocollapse */
    AyModule.ctorParameters = function () { return []; };
    return AyModule;
}());
export { AyModule };
