// @ay/ui/dist/table/table.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { TableComponent } from './table.component';
import { ColumnComponent } from './column.component';
import { ScrollModule } from "../scroll/scroll.module";
import { InputModule } from "../input/input.module";
import { ShowModule } from "../show/show.module";
import { FormatModule } from "../format/format.module";
export { TableComponent, ColumnComponent };
var TableModule = (function () {
    function TableModule() {
    }
    TableModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, FormsModule, ScrollModule, InputModule, ShowModule, FormatModule],
                    exports: [TableComponent, ColumnComponent],
                    declarations: [TableComponent, ColumnComponent]
                },] },
    ];
    /** @nocollapse */
    TableModule.ctorParameters = function () { return []; };
    return TableModule;
}());
export { TableModule };
