// @ay/ui/dist/tooltip/placements.js
export var Direction = null;
var attachments = {
    top: { attachment: 'bottom center', offset: '10px 0', opposite: 'top center' },
    left: { attachment: 'middle right', offset: '0 10px', opposite: 'middle left' },
    right: { attachment: 'middle left', offset: '0 -10px', opposite: 'middle right' },
    bottom: { attachment: 'top center', offset: '-10px 0', opposite: 'bottom center' },
    lb: { attachment: 'top left', offset: '-10px 0', opposite: 'bottom left' },
    rb: { attachment: 'top right', offset: '-10px 0', opposite: 'bottom right' },
    rt: { attachment: 'bottom right', offset: '10px 0', opposite: 'top right' },
    lt: { attachment: 'bottom left', offset: '10px 0', opposite: 'top left' },
};
var PLACEMENTS = Object.keys(attachments).reduce(function (placements, direction) {
    var _a = attachments[direction], attachment = _a.attachment, offset = _a.offset, opposite = _a.opposite;
    var targetAttachment = opposite;
    placements[direction] = { opposite: opposite, attachment: attachment, targetAttachment: targetAttachment, offset: offset };
    return placements;
}, {});
export function placement(direction) {
    return PLACEMENTS[direction];
}
